### PR TITLE
feat: Virtual leaves should be hashed as valid block items

### PR DESCRIPTION
**Pull request by thenswan**
**Description**:
This PR:
- Renames `VirtualLeaf` to `StateItem` in the `virtual_map_state.proto`;
- Renames `VirtualMapKey` and `VirtualMapValue` to `StateKey` and `StateValue`, respectively;
- Changes `VirtualLeafBytes` hashing bytes to wrap key+value into a `OneOf`.

**Related issue(s)**:
Fixes: https://github.com/hiero-ledger/hiero-consensus-node/issues/20371

**Checklist**
- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, migration, state validation, etc.)

### CODE REVIEW

**Review by imalygin — DISMISSED** [review]

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java [hunk 1]
```diff
@@ -237,28 +239,45 @@ public void writeTo(final WritableSequentialData out) {
...
+
+        // `1 +` for 0x00 prefix
+        return 1 + ProtoWriterTools.sizeOfDelimited(FIELD_MERKLELEAF_STATEITEM, innerLen);
     }
 
     // Output size must be at least getSizeInBytesForHashing()
     public void writeToForHashing(final BufferedData out) {
         out.reset();
         assert out.remaining() >= getSizeInBytesForHashing();
+
```
Comment (jasperpotts) [inline platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java:260]: Be nice to add comment for why we are adding the 0x00 byte prefix. This is because we are adding the 0x00 prefix byte to all leaf hashes in the Hiero Merkle tree. So that there is clear guaranteed domain separation of hash space between leaves and internal nodes.

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java [hunk 2]
```diff
@@ -237,28 +239,45 @@ public void writeTo(final WritableSequentialData out) {
...
-        len += 1; // 0x00 prefix
         final Bytes kb = keyBytes();
-        len += ProtoWriterTools.sizeOfDelimited(FIELD_LEAFRECORD_KEY, Math.toIntExact(kb.length()));
+        final int keyLen = Math.toIntExact(kb.length());
+        int innerLen = ProtoWriterTools.sizeOfDelimited(FIELD_LEAFRECORD_KEY, keyLen);
+
         final Bytes vb = valueBytes();
-        if (vb != null) {
-            len += ProtoWriterTools.sizeOfDelimited(FIELD_LEAFRECORD_VALUE, Math.toIntExact(vb.length()));
+        final int valueLen = (vb != null) ? Math.toIntExact(vb.length()) : -1;
```
Comment (artemananiev) [inline platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java]: If valueBytes is not null, but empty, it won't be serialized. What about
```java
final Bytes vb = valueBytes();
if (vb != null) {
    final int valueLen = Math.toIntExact(vb.length());
    if (valueLen > 0) {
        innerLen += ProtoWriterTools.sizeOfDelimited(FIELD_LEAFRECORD_VALUE, valueLen);
    }
}
```
?

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java [hunk 3]
```diff
@@ -237,28 +239,45 @@ public void writeTo(final WritableSequentialData out) {
...
         out.writeByte((byte) 0x00);
+
         final Bytes kb = keyBytes();
-        ProtoWriterTools.writeDelimited(out, FIELD_LEAFRECORD_KEY, Math.toIntExact(kb.length()), kb::writeTo);
+        final int keyLen = Math.toIntExact(kb.length());
+
         final Bytes vb = valueBytes();
-        if (vb != null) {
-            ProtoWriterTools.writeDelimited(out, FIELD_LEAFRECORD_VALUE, Math.toIntExact(vb.length()), vb::writeTo);
+        final int valueLen = (vb != null) ? Math.toIntExact(vb.length()) : -1;
```
Comment (artemananiev) [inline platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java]: Same as my comment above in `getSizeInBytesForHashing()`

Comment (thenswan) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java]: Done

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java (hunk 2 above)
Comment (thenswan) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java]: Sounds good, done

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java (hunk 1 above)
Comment (thenswan) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java:260]: Done, thank you

**Review by artemananiev — DISMISSED** [review]

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java [hunk 4]
```diff
@@ -237,28 +239,53 @@ public void writeTo(final WritableSequentialData out) {
...
+
         final Bytes kb = keyBytes();
-        ProtoWriterTools.writeDelimited(out, FIELD_LEAFRECORD_KEY, Math.toIntExact(kb.length()), kb::writeTo);
+        final int keyLen = Math.toIntExact(kb.length());
+        int innerLen = ProtoWriterTools.sizeOfDelimited(FIELD_LEAFRECORD_KEY, keyLen);
+
         final Bytes vb = valueBytes();
         if (vb != null) {
-            ProtoWriterTools.writeDelimited(out, FIELD_LEAFRECORD_VALUE, Math.toIntExact(vb.length()), vb::writeTo);
+            final int valueLen = Math.toIntExact(vb.length());
```
Comment (jasperpotts) [inline platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java:271]: is this the right handling of null? we do we support null as a valid return from valueBytes() should it just return empty array? how does VirtualMap handle the difference between null values and default values as a zero length byte[] in protobuf is a valid object with all default values.

Comment (jasperpotts) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java:271]: Are we testing all the edge cases like null, default etc? Also do we test that written objects are readable with PBJ generated code from schema?

Comment (artemananiev) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java:271]: These are good questions... At the State API level, nulls are explicitly prohibited:

```java
    protected void putIntoDataSource(@NonNull K key, @NonNull V value) { ... }
```

However, VirtualMap accepts nulls just fine:

```java
    public <V> void put(@NonNull final Bytes key, @Nullable final V value, @Nullable final Codec<V> valueCodec) { ... }
    public void putBytes(@NonNull final Bytes keyBytes, @Nullable final Bytes valueBytes) { ... }
```

When a value is null, it isn't written at all. When it gets deserialized later, it's initialized to null, too. So far so good.

However, there is another edge case. When a value is not null, but it's a default PBJ object (protobuf record with all fields set to default values), the codec will encode it as an empty byte array rather than null. This is handled fine, as far as I can see:

* `getValueBytes()` returns null, if the value is null, but an empty Bytes object, if the value is a default record
* `getSizeInBytes()` distinguishes between null values and empty values
* `writeTo()` handles empty values fine, too, they are written as length-delimited fields with zero length
* `parseFrom()` uses `Bytes.EMPTY`, if the field is there, and its length is zero, rather than set the value to null

The only issue is this new method, `writeForHashing()`. The null check at line 272 should be preserved, but the length checks below at lines 274 and 283 should be removed. `getSizeInBytesForHashing()` should be updated accordingly

**Review by imalygin — DISMISSED** [review]

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java (hunk 4 above)
Comment (jsync-swirlds) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java:271]: The codec appears to be behaving correctly (based on your description) for set-but-default and not-set cases.

**Review by artemananiev — APPROVED** [review]

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java [hunk 5]
```diff
@@ -237,28 +239,47 @@ public void writeTo(final WritableSequentialData out) {
...
         }
-        return len;
+
+        // `1 +` for 0x00 prefix
+        return 1 + ProtoWriterTools.sizeOfDelimited(FIELD_MERKLELEAF_STATEITEM, innerLen);
     }
 
     // Output size must be at least getSizeInBytesForHashing()
     public void writeToForHashing(final BufferedData out) {
         out.reset();
```
Comment (jsync-swirlds) [inline platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java:258]: Is the reset necessary?
What if I'm writing multiple leaf entries to a single buffer for hashing?
Esample:
```java
    left.writeToForHashing(hashingDigest);
    right.writeToForHashing(hashingDigest);
    return new Hash(hashingDigest.digest(), Cryptography.DEFAULT_DIGEST_TYPE);
```

**Review by jsync-swirlds — APPROVED** [review]
One question, the rest looks correct.

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java (hunk 4 above)
Comment (thenswan) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java:271]: @artemananiev I updated as you proposed, thanks for that.
cc @jasperpotts

File: platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java (hunk 5 above)
Comment (artemananiev) [reply platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/datasource/VirtualLeafBytes.java:258]: No, leaves are hashed one by one. Then leaf hashes are combined into internal node hashes in separate hashing tasks, so MessageDigest objects are always different

**Review by artemananiev — APPROVED** [review]

**Review by imalygin — APPROVED** [review]

**Review by mxtartaglia-sl — APPROVED** [review]
